### ISSUE TITLE
Integration with AWS Codebuild?

### ISSUE THREAD

**Issue by antsankov** [lambci/lambci#73]
AWS recently announced [Codebuild](https://aws.amazon.com/documentation/codebuild/) at Re:Invent this year, any plans on integrating it into Lambci for greater builder flexibility? 

This would solve the problem of a 5 minute build limit, allow for more flexibility of build images without going to ECS, preserve statlessness, and allow provisioning more resources for a build.

**Comment by mhart** [comment]
@antsankov in what cases would you want to use LambCI if you're using CodeBuild?

**Comment by antsankov** [comment]
Specifically use cases surrounding the infrastructure around builds:

* Responding to Github [webhooks](http://stackoverflow.com/questions/40964349/is-it-possible-to-combine-aws-codebuild-and-codepipeline-to-build-described-ci-w).
* Merge behavior - Amazon doesn't make it easy for you to customize merge and branch behavior.
* Notifying subscribers of build events - Github badge, Slack bot, etc. 
* Cheaper than Codepipeline - This service does cost money per active pipeline, which can add up if you're maintaining a separate pipeline for forks of a repo.

**Comment by mhart** [comment]
Hmmm, interesting – so mostly cases where you don't want to use CodePipeline then? (I thought that supported GitHub?)

I'll leave this open – I kinda feel like most of deficiencies in CodeBuild/CodePipeline will just be fixed in time by AWS – but I guess it wouldn't be bad to explore this as an option for LambCI in the meantime.

**Comment by antsankov** [comment]
Definitely. I'm fascinated by the promise of completely stateless CI, and am curious about how this project evolves in the future.

**Comment by lielran** [comment]
I thought about it too. The integration with PR/Slack/... Is indeed missing in code build.
I guess the(my) main pain points are 5min limit and 500mb disk space.

I thought about integrated with Function step(also from the last reinvent) or even poc GCE functions

**Comment by voidfiles** [comment]
Hi,

First, I wanted to say that LambCI is an interesting project, and love all the work that has gone into it so far.

Second, I think this is a good idea as well.

I might be missing something, and if I am please let me know, but I think you might be able to use CodeBuild almost like you would use an ECS task, except you don't need to manage an ECS pool, and you can choose an docker image as your base build image.

This is interesting to me for a couple of reasons.

Mainly, Many CI pipelines don't have first class support for Github PR's, and I'd like to have fine tune control over that flow. Like, I want to run static analysis, and unit-tests, but I don't want to run functional tests or deploy to an integration environment.
 
Regarding CodeBuild, I think you might be able to initiate a CodeBuild, and then subscribe to an SNS for when it is done.

I don't know if I could commit to doing this work, but if I am in the ballpark I wouldn't mind trying something.

**Comment by jimsheldon** [comment]
I would also be interested in whether CodeBuild could replace ECS in LambCI

**Comment by lorengordon** [comment]
> I might be missing something, and if I am please let me know, but I think you might be able to use CodeBuild almost like you would use an ECS task, except you don't need to manage an ECS pool, and you can choose an docker image as your base build image.

This. Yes.

I find it interesting that the [`StartBuild` API](https://docs.aws.amazon.com/codebuild/latest/APIReference/API_StartBuild.html) can override what looks like every parameter of a CodeBuild project, so depending on the desired abstraction, a single CodeBuild Project could maybe be used for all lambci jobs, with the event details and the lambci config determining the params to send along to the StartBuild request.

**Comment by reegnz** [comment]
Jenkins has a plugin to use CodeBuild as an agent:
https://docs.aws.amazon.com/codebuild/latest/userguide/jenkins-plugin.html
https://github.com/jenkinsci/aws-codebuild-plugin

That could serve as inspiration on how to do it, since I'm guessing the API invocation is kind-of similar.
